File: formula-java/src/main/java/org/formula/parse/shuntingyard/Term.java
package org.formula.parse.shuntingyard;

import java.util.Optional;
import org.formula.Resolvable;
import org.formula.ResolvedValue;
import org.formula.context.DataContext;

public record Term(ResolvedValue value) implements Resolvable, Node {

    public static Term of(ResolvedValue value) {
        return new Term(value);
    }

    public static Term of(String value) {
        return new Term(ResolvedValue.of(value));
    }

    public static Term of(int value) {
        return new Term(ResolvedValue.of(value));
    }

    public static Term of(double value) {
        return new Term(ResolvedValue.of(value));
    }

    public static Term of(boolean value) {
        return new Term(ResolvedValue.of(value));
    }

    @Override
    public ResolvedValue resolve(DataContext context) {
        return Optional.ofNullable(value)
                .orElse(ResolvedValue.none());
    }
}

File: formula-cs/Formula/ShuntingYard/Term.cs
namespace Formula.ShuntingYard;

public readonly record struct Term(ResolvedValue Value, string Prefix = "", string Suffix = "") : IResolvable, INode
{
    public ResolvedValue Resolve(IDataContext context)
    {
        return Prefix.Length > 0 || Suffix.Length > 0
            ? QuotedTextResolvedValue.Of(Value, Prefix, Suffix)
            : Value;
        // return Optional.ofNullable(value)
        //     .map(value -> prefix != null && suffix != null
        //         ? QuotedTextResolvedValue.of(value, prefix, suffix)
        //         : value)
        //     .orElse(ResolvedValue.none());
    }
}